library/src/main/java/org/apache/apex/malhar/lib/wal/FileSystemWAL.java

```diff
@@ -0,0 +1,688 @@
+ * under the License.
+ */
+package org.apache.apex.malhar.lib.wal;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+import javax.validation.constraints.Min;
+import javax.validation.constraints.NotNull;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.apex.malhar.lib.utils.FileContextUtils;
+import org.apache.apex.malhar.lib.utils.IOUtils;
+import org.apache.hadoop.fs.CreateFlag;
+import org.apache.hadoop.fs.FileContext;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.Options;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.RemoteIterator;
+import org.apache.hadoop.fs.Syncable;
+import org.apache.hadoop.fs.local.LocalFs;
+import org.apache.hadoop.fs.local.RawLocalFs;
+
+import com.google.common.base.Preconditions;
+
+import com.datatorrent.api.annotation.Stateless;
+import com.datatorrent.netlet.util.Slice;
+
+/**
+ * A WAL implementation that is file based.
+ * <p/>
+ * Note:<br/>
+ * The FileSystem Writer and Reader operations should not alternate because intermingling these operations will cause
+ * problems. Typically the  WAL Reader will only used in recovery.<br/>
+ *
+ * Also this implementation is thread unsafe- the filesystem wal writer and reader operations should be performed in
+ * operator's thread.
+ */
+public class FileSystemWAL implements WAL<FileSystemWAL.FileSystemWALReader, FileSystemWAL.FileSystemWALWriter>
+{
+
+  @NotNull
+  private String filePath;
+
+  //max length of the file
+  @Min(0)
+  private long maxLength;
+
+  private FileSystemWALPointer walStartPointer = new FileSystemWALPointer(0, 0);
+
+  @NotNull
+  private FileSystemWAL.FileSystemWALReader fileSystemWALReader = new FileSystemWALReader(this);
+
+  @NotNull
+  private FileSystemWAL.FileSystemWALWriter fileSystemWALWriter = new FileSystemWALWriter(this);
+
+  //part => tmp file path;
+  private final Map<Integer, String> tempPartFiles = new TreeMap<>();
+
+  private long lastCheckpointedWindow = Stateless.WINDOW_ID;
+
+  @Override
+  public void setup()
+  {
+    try {
+      FileContext fileContext = FileContextUtils.getFileContext(filePath);
+      if (maxLength == 0) {
+        maxLength = fileContext.getDefaultFileSystem().getServerDefaults().getBlockSize();
+      }
+      fileSystemWALWriter.open(fileContext);
+      fileSystemWALReader.open(fileContext);
+    } catch (IOException e) {
+      throw new RuntimeException("during setup", e);
+    }
+  }
+
+  @Override
+  public void beforeCheckpoint(long window)
+  {
+    try {
+      lastCheckpointedWindow = window;
+      fileSystemWALWriter.flush();
+    } catch (IOException e) {
+      throw new RuntimeException("during before cp", e);
+    }
+  }
+
+  @Override
+  public void committed(long window)
```

Comment:
Ok I will.
This is HDFS issue with the rename+overwrite option that I was referring to
https://issues.apache.org/jira/browse/HDFS-6757
It was advised to avoid  rename + overwrite option on under construction file.

Should I document all this as well here?